src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -227,30 +462,128 @@ Gradient(const arma::mat& /*parameters*/, const size_t i, arma::mat& gradient,
   }
 
   counter++;
+  currentBatch++;
 
-  if (counter >= numFunctions)
+  if (counter * batchSize >= numFunctions)
   {
     counter = 0;
-    currentBatch++;
   }
-  else if (counter % batchSize == 0)
+
+  if (preTrainSize > 0)
+  {
+    preTrainSize--;
+  }
+}
+
+template<
+  typename Model,
+  typename InitializationRuleType,
+  typename Noise,
+  typename PolicyType
+>
+template<typename Policy>
+typename std::enable_if<std::is_same<Policy, WGANGP>::value,
+                        void>::type
+GAN<Model, InitializationRuleType, Noise, PolicyType>::
+Gradient(const arma::mat& /* parameters */,
+         const size_t i,
+         arma::mat& gradient,
+         const size_t /* batchSize */)
+{
+  if (!reset)
+    Reset();
+
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+      Reset();
+    gradient = arma::zeros<arma::mat>(parameter.n_elem, 1);
+  }
+  else
+    gradient.zeros();
+
+  if (noiseGradientDiscriminator.is_empty())
+  {
+    noiseGradientDiscriminator = arma::zeros<arma::mat>(
+        gradientDiscriminator.n_elem, 1);
+  }
+  else
+  {
+    noiseGradientDiscriminator.zeros();
+  }
+
+  gradientGenerator = arma::mat(gradient.memptr(),
+      generator.Parameters().n_elem, 1, false, false);
+
+  gradientDiscriminator = arma::mat(gradient.memptr() +
+      gradientGenerator.n_elem,
+      discriminator.Parameters().n_elem, 1, false, false);
+
+  // Get the gradients of the Discriminator.
+  discriminator.Gradient(discriminator.parameter, i, gradientDiscriminator,
+      batchSize);
+  noise.imbue( [&]() { return noiseFunction();} );
+  generator.Forward(std::move(noise));
+  discriminator.predictors.cols(numFunctions, numFunctions + batchSize - 1) =
+      boost::apply_visitor(outputParameterVisitor, generator.network.back());
+
+  discriminator.responses.cols(numFunctions, numFunctions + batchSize - 1) =
+      -arma::ones(1, batchSize);
+  discriminator.Gradient(discriminator.parameter, numFunctions,
+      noiseGradientDiscriminator, batchSize);
+  gradientDiscriminator += noiseGradientDiscriminator;
+
+  if (currentBatch % generatorUpdateStep == 0 && preTrainSize == 0)
+  {
+    // Minimize -D(G(noise)).
+    // Pass the error from Discriminator to Generator.
+    discriminator.responses.cols(numFunctions, numFunctions + batchSize - 1) =
+        arma::ones(1, batchSize);
+    discriminator.Gradient(discriminator.parameter, numFunctions,
+        noiseGradientDiscriminator, batchSize);
+    generator.error = boost::apply_visitor(deltaVisitor,
+        discriminator.network[1]);
+
+    generator.Predictors() = noise;
+    generator.ResetGradients(gradientGenerator);
+    generator.Gradient(generator.parameter, 0, gradientGenerator, batchSize);
+
+    gradientGenerator *= multiplier;
+  }
+
+  counter++;
+  currentBatch++;
+
+  if (counter * batchSize >= numFunctions)
```

Comment:
Might be a good idea to add a comment here, that this is used to control the pretrain step.